components/script_plugins/webidl_must_inherit.rs

```diff
@@ -0,0 +1,223 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use rustc::hir::{self, HirId};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintContext, LintPass};
+use rustc::ty;
+use std::boxed;
+use std::env;
+use std::error::Error;
+use std::fmt;
+use std::fs;
+use std::io;
+use std::path;
+use syntax::ast;
+use webidl::ast::*;
+use webidl::visitor::*;
+use webidl::*;
+
+declare_lint!(
+    WEBIDL_INHERIT_CORRECT,
+    Deny,
+    "Warn and report usage of incorrect webidl inheritance"
+);
+
+pub struct WebIdlPass;
+
+#[derive(Clone, Debug)]
+pub enum WebIdlError {
+    ParentMismatch {
+        name: String,
+        rust_parent: String,
+        webidl_parent: String,
+    },
+}
+
+impl fmt::Display for WebIdlError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            WebIdlError::ParentMismatch {
+                name,
+                rust_parent,
+                webidl_parent,
+            } => {
+                return write!(f, "webidl-rust inheritance mismatch, rust: {:?}, rust parent: {:?}, webidl parent: {:?}",
+                    &name, &rust_parent, &webidl_parent);
+            },
+        }
+    }
+}
+
+impl Error for WebIdlError {
+    fn description(&self) -> &str {
+        "WebIdlError"
+    }
+
+    fn cause(&self) -> Option<&Error> {
+        None
+    }
+}
+
+impl WebIdlPass {
+    pub fn new() -> WebIdlPass {
+        WebIdlPass
+    }
+}
+
+fn get_typ_name(typ: String) -> String {
```

Comment:
* Since `type` is a keyword, by convention we usually shorten it to `ty`.
* Is `get_typ_name` a typo?
* Can this function receive a `&str` instead of a `String`?

Reply:
It is not a typo, will change to `ty` and `get_ty_name`.